### PR TITLE
Supervision Job for Active Failover

### PR DESCRIPTION
**Pull request by graetzer**
- Support for new master selection in Supervision
- Enables support for multiple slaves

Design document:
https://github.com/arangodb/documents/blob/master/FeatureProposals/ActiveFailoverJob.md

PR-Test:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/295/

### CODE REVIEW

File: arangod/Agency/AgencyFeature.cpp [hunk 1]
```diff
@@ -245,7 +245,7 @@ void AgencyFeature::start() {
   // Find the agency prefix:
   auto feature = ApplicationServer::getFeature<ClusterFeature>("Cluster");
   if (!feature->agencyPrefix().empty()) {
-    arangodb::consensus::Supervision::setAgencyPrefix(
+    consensus::Supervision::setAgencyPrefix(
```
Comment (kvahed) [inline arangod/Agency/AgencyFeature.cpp]: I never approve just consmetic changes :). There is really no necessity. It also only mangles git blame.

File: arangod/Agency/Job.h [hunk 2]
```diff
@@ -263,8 +267,7 @@ inline arangodb::consensus::trans_ret_t generalTransaction(
 }
 
 inline arangodb::consensus::trans_ret_t transient(AgentInterface* _agent,
-                                                  Builder const& transaction,
-                                                  bool waitForCommit = true) {
```
Comment (kvahed) [inline arangod/Agency/Job.h]: Consider setting the default for waitForSync in the header. Wouldn't you say that it helps readability?

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 3]
```diff
@@ -953,13 +953,13 @@ Result RestReplicationHandler::processRestoreCollection(
   ExecContext const* exe = ExecContext::CURRENT;
   if (name[0] != '_' && exe != nullptr && !exe->isSuperuser() &&
       ServerState::instance()->isSingleServer()) {
-    AuthenticationFeature* af = AuthenticationFeature::instance();
```
Comment (kvahed) [inline arangod/RestHandler/RestReplicationHandler.cpp]: Cosmetic again. Is this a race for who touches most lines of code in a week?

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 4]
```diff
@@ -1124,12 +1124,14 @@ Result RestReplicationHandler::processRestoreCollectionCoordinator(
 
     ExecContext const* exe = ExecContext::CURRENT;
     if (name[0] != '_' && exe != nullptr && !exe->isSuperuser()) {
-      AuthenticationFeature* af = AuthenticationFeature::instance();
```
Comment (kvahed) [inline arangod/RestHandler/RestReplicationHandler.cpp]: Same as above

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 5]
```diff
@@ -1325,7 +1327,9 @@ Result RestReplicationHandler::processRestoreUsersBatch(
   auto queryResult = query.execute(queryRegistry);
 
   AuthenticationFeature* af = AuthenticationFeature::instance();
-  af->userManager()->outdate();
+  if (af->userManager() != nullptr) {
```
Comment (kvahed) [inline arangod/RestHandler/RestReplicationHandler.cpp]: Not the same as before. Consider `TRI_ASSERT `in `MAINTAINER_MODE`

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 6]
```diff
@@ -1124,12 +1124,14 @@ Result RestReplicationHandler::processRestoreCollectionCoordinator(
...
     ExecContext const* exe = ExecContext::CURRENT;
     if (name[0] != '_' && exe != nullptr && !exe->isSuperuser()) {
-      AuthenticationFeature* af = AuthenticationFeature::instance();
-      af->userManager()->updateUser(ExecContext::CURRENT->user(),
-                   [&](auth::User& entry) {
-                     entry.grantCollection(dbName, col->name(), auth::Level::RW);
-                     return TRI_ERROR_NO_ERROR;
-                   });
+      auth::UserManager* um = AuthenticationFeature::instance()->userManager();
+      if (um != nullptr) {
```
Comment (kvahed) [inline arangod/RestHandler/RestReplicationHandler.cpp]: Consider adding assertion in `MAINTAINER_MODE`

**Comment by graetzer** [comment]
The design document is linked in the PR description.

File: arangod/Agency/Job.h (hunk 2 above)
Comment (graetzer) [reply arangod/Agency/Job.h]: wait for commit was not used here

File: arangod/Agency/ActiveFailoverJob.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        _jb->add("jobId", VPackValue(_jobId));
+        _jb->add("timeCreated", VPackValue(now));
+      } // todo
+      
+      // FailedServers entry []
+      _jb->add(VPackValue(failedServersPrefix + "/" + _server));
+      { VPackArrayBuilder failedServers(_jb.get()); }
+    } // Operations
+    
+      //Preconditions
```
Comment (neunhoef) [inline arangod/Agency/ActiveFailoverJob.cpp]: Wrong indentation.

File: js/client/tests/active-failover/basic.js [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+      bearer: jwtRoot,
+    }
+  });
+
+
+  assertTrue(res instanceof request.Response);
+  //assertTrue(res.hasOwnProperty('statusCode'));
+  assertTrue(res.statusCode === 200);
+  //assertTrue(res.hasOwnProperty('json'));
+  //print ("sdsdssddata 2", res.json);
```
Comment (neunhoef) [inline js/client/tests/active-failover/basic.js]: We should remove this.

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/328/

**Pull request merged**